csep/utils/csep1_unit_tests

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Mon Jan 27 14:38:30 2020
+Implementation of Unit Test for L Test
+@author: khawaja
+"""
+import numpy
+import unittest
+from csep1_tests import n_test
+from csep1_tests import poisson_log_likelihood
+from csep1_tests import _forecast_realization_inverse_cdf
+from csep1_tests import _m_test_prep
+from csep1_tests import _s_test_prep
+from csep1_tests import _conditional_l_test_prep
+
+
+class TestGriddedForecastTests(unittest.TestCase):
+    
+
+    def test_n_test(self):
+        forecast =numpy.zeros((10,10))+0.0015
+        forecast /= numpy.size(forecast)
+        observation = numpy.zeros((10,10))
+        expected_output = {'delta1':1.0,'delta2':0.9985011244377109,'model_name':'Unittesting'}
+        print('N Test: Running Unit Test')
+        self.assertEqual(n_test(observation, forecast,'Unittesting'), expected_output, 'Failed N Test')
+        
+    def test_l_lest(self):
+        
+        """
+        L Test is bisected in 03 independent Steps. 
+        #Step 1: Test the Likelihood between Actual Observation and Actual Forecast
+        #Step 2: Test the Generation of Realization of Forecast by providing Random Matrix
+        #Step 3: Test applied to check the Quantile Score on Eq. 15 of Zechar et al, 2010 
+        """
+        observation = numpy.array([[5, 8],[4, 2]])
+        forecast = numpy.array([[8, 2],[3,5]])
+        random_matrix = numpy.array([[0.2, 0.4],[0.6, 0.8]])
+        num_simulations = 1
+        Actual_log_Likelihood = numpy.array([[-2.3902840343828666,-7.059425458265689],
+                              [-1.7836046756755066,-2.474271355691745]])
+        ll_actual=-13.707585524015807
+        forecast_realization = numpy.array([[6., 1.],[3., 7.]])
+        ll_simulations = -7.16447335862158
+        Quantile_score = 0
+        
+        print('L Test: Step 1, Running NUMPY Test')
+        numpy.testing.assert_array_equal(poisson_log_likelihood(observation, forecast), Actual_log_Likelihood,'Failed L Test: Actual Likelihood Testing Failed')
+        
+        print('L Test: Step 2, Running NUMPY Test')
+        numpy.testing.assert_array_equal(_forecast_realization_inverse_cdf(random_matrix, forecast), forecast_realization, 'Failed L Test: Simulated Forecast Likelihood Test')
+        
+        print('L Test: Step 3, Running Unit Test')
+        self.assertEqual((numpy.sum(ll_simulations <= ll_actual) / num_simulations), Quantile_score,'Failed L Test: Quantile Score Testing')
+      
+        
+    def test_m_test_partial(self):
```

Comment:
im not sure these tests actually test the function itself. we want to call the function with the inputs and check that the outputs are what is expected.

these tests are looking to make sure indivdual parts of the function work but not the overall component.

Reply:
Yes, you are right. When you were here in Potsdam and we started writing likelihood based tests (L test at that time) and we discussed this thing. Likelihood based tests use some random numbers generation inside them for creating forecasts realizations. So there were two options for us to write Unit tests. Either we pass a pre-selected random numbers for every simulation as an input to test the working of the rest of likelihood test function. Or the 2nd (clever / easy) option was to test each component of the function, separately. We discussed that if, all the parts are individually giving correct results, then they should also be working fine together. Therefore, unit tests for L, M, S and CL tests are based on testing of individual components.